extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java

```diff
@@ -145,28 +152,30 @@ private void processCloudEvent(RoutingContext routingContext) {
                     log.warnf("Unexpected CloudEvent spec-version '%s'.", ceSpecVersion);
                 }
 
-                final FunctionInvoker invoker;
+                Collection<FunctionInvoker> candidates = new ArrayList<>();
                 if (defaultInvoker != null) {
-                    invoker = defaultInvoker;
+                    candidates.add(defaultInvoker);
                 } else {
-                    invoker = typeTriggers.get(ceType);
-                    if (invoker == null) {
+                    candidates = typeTriggers.get(ceType);
+
+                    if (candidates == null || candidates.isEmpty()) {
                         routingContext.fail(404);
                         log.error("Couldn't map CloudEvent type: '" + ceType + "' to a function.");
                         return;
                     }
                 }
 
-                final Type inputCeDataType = (Type) invoker.getBindingContext().get(INPUT_CE_DATA_TYPE);
-                final Type outputCeDataType = (Type) invoker.getBindingContext().get(OUTPUT_CE_DATA_TYPE);
-                final Type innerInputType = inputCeDataType != null ? inputCeDataType : invoker.getInputType();
-                final Type innerOutputType = outputCeDataType != null ? outputCeDataType : invoker.getOutputType();
-                final ObjectReader reader = (ObjectReader) invoker.getBindingContext().get(DATA_OBJECT_READER);
-                final ObjectWriter writer = (ObjectWriter) invoker.getBindingContext().get(DATA_OBJECT_WRITER);
+                boolean anyMatched = false;
+                for (FunctionInvoker invoker : candidates) {
```

Comment:
I don't think this all should be in for loop.

We should first check how many matching candidates are there.
zero -- return `404`
one -- execute the single matching candidate
two or more -- return `409`

Reply:
hmm looking more carefully it might be difficult to get rid of the loop, maybe leave it as it is

Reply:
I had the same round of tries and ended up with the loop :) so let's keep the loop and I think it should return directly after first matching invoker, agreed?

Reply:
>so let's keep the loop and I think it should return directly after first matching invoker, agreed?

I don't like the idea. It would pick single function basically at random, I think it depends on order in which annotations are processed. There should be just one match and if there are multiple we should return `409`.

What about this:

diff against upstream not your branch:
```diff
diff --git a/extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java b/extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java
index 5de658751d..a584cdc3c7 100644
--- a/extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java
+++ b/extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/runtime/bindings/knative/events/VertxRequestHandler.java
@@ -12,11 +12,11 @@ import static io.quarkus.funqy.runtime.bindings.knative.events.KnativeEventsBind
 import java.io.IOException;
 import java.io.InputStream;
 import java.lang.reflect.Type;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.UUID;
+import java.util.*;
 import java.util.concurrent.Executor;
 import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 import javax.enterprise.inject.Instance;
 import javax.enterprise.inject.spi.CDI;
@@ -61,7 +61,8 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
     protected final CurrentVertxRequest currentVertxRequest;
     protected final Executor executor;
     protected final FunctionInvoker defaultInvoker;
-    protected final Map<String, FunctionInvoker> typeTriggers;
+    protected final Map<String, Collection<FunctionInvoker>> typeTriggers;
+    protected final Map<String, List<Predicate<CloudEvent>>> invokersFilters;
     protected final String rootPath;
 
     public VertxRequestHandler(Vertx vertx,
@@ -70,7 +71,8 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
             ObjectMapper mapper,
             FunqyKnativeEventsConfig config,
             FunctionInvoker defaultInvoker,
-            Map<String, FunctionInvoker> typeTriggers,
+            Map<String, Collection<FunctionInvoker>> typeTriggers,
+            Map<String, List<Predicate<CloudEvent>>> invokersFilters,
             Executor executor) {
         this.rootPath = rootPath;
         this.defaultInvoker = defaultInvoker;
@@ -79,6 +81,7 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
         this.executor = executor;
         this.mapper = mapper;
         this.typeTriggers = typeTriggers;
+        this.invokersFilters = invokersFilters;
         Instance<CurrentIdentityAssociation> association = CDI.current().select(CurrentIdentityAssociation.class);
         this.association = association.isResolvable() ? association.get() : null;
         this.currentVertxRequest = CDI.current().select(CurrentVertxRequest.class).get();
@@ -145,18 +148,52 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
                     log.warnf("Unexpected CloudEvent spec-version '%s'.", ceSpecVersion);
                 }
 
-                final FunctionInvoker invoker;
+                final Collection<FunctionInvoker> candidates;
                 if (defaultInvoker != null) {
-                    invoker = defaultInvoker;
+                    candidates = Collections.singleton(defaultInvoker);
                 } else {
-                    invoker = typeTriggers.get(ceType);
-                    if (invoker == null) {
+                    candidates = typeTriggers.get(ceType);
+
+                    if (candidates == null || candidates.isEmpty()) {
                         routingContext.fail(404);
                         log.error("Couldn't map CloudEvent type: '" + ceType + "' to a function.");
                         return;
                     }
                 }
 
+                final CloudEvent<?> evt;
+                if (binaryCE) {
+                    evt = new HeaderCloudEventImpl<>(
+                            httpRequest.headers(),
+                            null,
+                            null,
+                            null,
+                            null);
+                } else {
+                    evt = new JsonCloudEventImpl<>(
+                            structuredPayload,
+                            null,
+                            null,
+                            null);
+                }
+                List<FunctionInvoker> matchingCandidates = candidates
+                        .stream()
+                        .filter(fi -> match(fi, evt))
+                        .collect(Collectors.toList());
+
+                if (matchingCandidates.size() <= 0) {
+                    routingContext.fail(404);
+                    log.error("Couldn't map CloudEvent type: '" + ceType + "' to any function.");
+                    return;
+                }
+                if (matchingCandidates.size() > 1) {
+                    routingContext.fail(409);
+                    log.error("CloudEvent type: '" + ceType + "' matches multiple function.");
+                    return;
+                }
+
+                final FunctionInvoker invoker = matchingCandidates.get(0);
+
                 final Type inputCeDataType = (Type) invoker.getBindingContext().get(INPUT_CE_DATA_TYPE);
                 final Type outputCeDataType = (Type) invoker.getBindingContext().get(OUTPUT_CE_DATA_TYPE);
                 final Type innerInputType = inputCeDataType != null ? inputCeDataType : invoker.getInputType();
@@ -165,6 +202,7 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
                 final ObjectWriter writer = (ObjectWriter) invoker.getBindingContext().get(DATA_OBJECT_WRITER);
 
                 final CloudEvent<?> inputCloudEvent;
+
                 final Object input;
                 if (invoker.hasInput()) {
                     if (binaryCE) {
@@ -363,6 +401,15 @@ public class VertxRequestHandler implements Handler<RoutingContext> {
 
     }
 
+    private boolean match(FunctionInvoker invoker, CloudEvent<?> inputCloudEvent) {
+
+        if (invokersFilters.get(invoker.getName()) == null || invokersFilters.get(invoker.getName()).isEmpty()) {
+            return true;
+        }
+
+        return invokersFilters.get(invoker.getName()).stream().allMatch(p -> p.test(inputCloudEvent));
+    }
+
     private void regularFunqyHttp(RoutingContext routingContext) {
         String path = routingContext.request().path();
         if (path == null) {
```

Reply:
so we create the cloud events with basic info required to evaluate filters and then build them again fully populated, did I got it right?

Reply:
yes, that's the idea, we just need attributes other that `data`.